### PR TITLE
Allow msgpack {de,}serialization of Time and AS::TimeWithZone

### PR DESCRIPTION
**Pull request by adamruzicka**

### CODE REVIEW

File: lib/dynflow/extensions/msgpack.rb [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+        end
+      end
+
+      ::Time.include ::Dynflow::Extensions::MsgPack::Time
+      ::MessagePack::DefaultFactory.register_type(0x00, Time, packer: MessagePack::Time::Packer, unpacker: MessagePack::Time::Unpacker)
+
+      begin
+        require 'active_support/time_with_zone'
+        ::ActiveSupport::TimeWithZone.include ::Dynflow::Extensions::MsgPack::Time
+        ::MessagePack::DefaultFactory.register_type(0x01, ActiveSupport::TimeWithZone, packer: MessagePack::Time::Packer, unpacker: MessagePack::Time::Unpacker)
```
Comment (adamruzicka) [inline lib/dynflow/extensions/msgpack.rb:20]: This is slightly wonky. Once `ActiveSupport::TimeWithZone` instance is serialized and then deserialized, it becomes just `Time`, but still `==` to the original instance. Still better then what we had before

Comment (adamruzicka) [reply lib/dynflow/extensions/msgpack.rb]: Not anymore

**Review by ezr-ondrej — APPROVED** [review]

**Comment by jlsherrill** [comment]
with this i'm getting:
```
Actions::Pulp3::AnsibleCollectionSyncTest#test_sync_mirror_true:
NoMethodError: undefined method `to_hash' for Wed, 18 Aug 2021 20:22:28 +0000:DateTime
Did you mean?  to_s
    foreman-tasks (5.1.1) lib/foreman_tasks.rb:20:in `trigger'
    foreman-tasks (5.1.1) lib/foreman_tasks.rb:26:in `block in trigger_task'
    foreman-tasks (5.1.1) lib/foreman_tasks.rb:46:in `block in rails_safe_trigger_task'
    foreman-tasks (5.1.1) lib/foreman_tasks.rb:45:in `rails_safe_trigger_task'
    foreman-tasks (5.1.1) lib/foreman_tasks.rb:24:in `trigger_task'
    foreman-tasks (5.1.1) lib/foreman_tasks.rb:55:in `sync_task'
    /home/vagrant/git/katello/test/support/pulp3_support.rb:70:in `create_repo'
    ./katello/test/actions/pulp3/orchestration/ansible_collection_sync_test.rb:11:in `setup'
    /home/vagrant/git/katello/test/support/vcr.rb:26:in `block in run'
    /home/vagrant/git/katello/test/support/vcr.rb:25:in `run'

```

**Comment by jlsherrill** [comment]
and test.log doesn't provide any details (in fact the traceback doesn't appear there)

**Comment by jlsherrill** [comment]
i added a debugger and grabbed some more output:
```
/home/vagrant/git/dynflow/lib/dynflow/serializable.rb:55:in `recursive_to_hash'
/home/vagrant/git/dynflow/lib/dynflow/serializable.rb:51:in `block in recursive_to_hash'
/home/vagrant/git/dynflow/lib/dynflow/serializable.rb:51:in `each'
/home/vagrant/git/dynflow/lib/dynflow/serializable.rb:51:in `inject'
/home/vagrant/git/dynflow/lib/dynflow/serializable.rb:51:in `recursive_to_hash'
/home/vagrant/git/dynflow/lib/dynflow/serializable.rb:51:in `block in recursive_to_hash'
/home/vagrant/git/dynflow/lib/dynflow/serializable.rb:51:in `each'
/home/vagrant/git/dynflow/lib/dynflow/serializable.rb:51:in `inject'
/home/vagrant/git/dynflow/lib/dynflow/serializable.rb:51:in `recursive_to_hash'
/home/vagrant/git/dynflow/lib/dynflow/serializable.rb:51:in `block in recursive_to_hash'
/home/vagrant/git/dynflow/lib/dynflow/serializable.rb:51:in `each'
/home/vagrant/git/dynflow/lib/dynflow/serializable.rb:51:in `inject'
/home/vagrant/git/dynflow/lib/dynflow/serializable.rb:51:in `recursive_to_hash'
/home/vagrant/git/dynflow/lib/dynflow/serializable.rb:59:in `recursive_to_hash'
/home/vagrant/git/dynflow/lib/dynflow/action.rb:263:in `to_hash'
/home/vagrant/git/dynflow/lib/dynflow/persistence.rb:46:in `save_action'
/home/vagrant/git/dynflow/lib/dynflow/execution_plan/steps/abstract_flow_step.rb:33:in `open_action'
/home/vagrant/git/dynflow/lib/dynflow/execution_plan/steps/abstract_flow_step.rb:16:in `execute'
/home/vagrant/git/dynflow/lib/dynflow/director.rb:69:in `execute'
/home/vagrant/git/dynflow/lib/dynflow/testing/in_thread_executor.rb:31:in `handle_work'
/home/vagrant/git/dynflow/lib/dynflow/testing/in_thread_executor.rb:19:in `process_work_items'
/home/vagrant/git/dynflow/lib/dynflow/testing/in_thread_executor.rb:13:in `execute'
/home/vagrant/git/dynflow/lib/dynflow/testing/in_thread_world.rb:48:in `execute'
/home/vagrant/git/dynflow/lib/dynflow/world.rb:185:in `trigger'
```

**Comment by adamruzicka** [comment]
> i added a debugger and grabbed some more output

I'm getting the same when trying to pass a datetime as actions input/output even on a pre-msgpack dynflow. Is it possible that there's something else that changed somewhere else?

**Comment by jlsherrill** [comment]
the tests seem to run fine against dynflow 1.5.0 

let me try your reproducer

**Comment by jlsherrill** [comment]
Actually the reproducer fails within the rails console for me:

```
 ForemanTasks.dynflow.world.trigger(Foo, DateTime.current)
2021-09-22T15:41:21 [E|bac|] undefined method `to_hash' for Wed, 22 Sep 2021 17:41:21 +0200:DateTime
 | Did you mean?  to_s (NoMethodError)

```

**Comment by adamruzicka** [comment]
@jlsherrill the latest commit should fix it, although I don't have a good feeling about it

**Comment by adamruzicka** [comment]
Let's hold this for a bit, I think I may be on to something

**Comment by jlsherrill** [comment]
Just to recap, this does appear to fix it for me.  So i'm good with this, but you seem to have some concerns.

**Comment by ezr-ondrej** [comment]
well, the when clause is not great but it fixes the issue :shrug:

**Pull request merged**